language/move-lang/src/typing/translate.rs

```diff
@@ -570,8 +572,66 @@ fn struct_def(context: &mut Context, s: &mut N::StructDefinition) {
     }
     core::solve_constraints(context);
 
+    let type_params: HashMap<_, _> = s
+        .type_parameters
+        .iter()
+        .map(|sp| {
+            let loc = sp.param.user_specified_name.loc;
+            (sp.param.id, (sp.is_phantom, loc))
+        })
+        .collect();
+
     for (_field_loc, _field_, idx_ty) in field_map.iter_mut() {
         expand::type_(context, &mut idx_ty.1);
+        check_phantom_params(context, &type_params, &idx_ty.1, false);
+    }
+}
+
+fn check_phantom_params(
+    context: &mut Context,
+    type_parameters: &HashMap<TParamID, (bool, Loc)>,
+    ty: &Type,
+    is_phantom_pos: bool,
+) {
+    match &ty.value {
+        Type_::Ref(_, inner) => check_phantom_params(context, type_parameters, &inner, false),
+        Type_::Param(param) => {
+            // Don't bother checking unbound type parameters as it should fail somewhere else
+            if let Some(&(is_phantom, param_loc)) = type_parameters.get(&param.id) {
+                if is_phantom && !is_phantom_pos {
+                    let msg = format!(
+                        "Phantom type parameter '{}' used in non-phantom position",
+                        param.user_specified_name
+                    );
+                    context.env.add_diag(diag!(
+                        Declarations::InvalidPhantomUse,
+                        (ty.loc, msg),
+                        (param_loc, "Type parameter declared here as phantom")
```

Comment:
One possible thought, instead of `is_phantom_pos` do something like `non_phantom_pos: Option<TypeArgument | FieldType>` And have it start with `Some(FieldType)`

Then you can check `non_phantom_pos.is_some()` and either 
- "Phantom type parameters cannot be used as the type for fields" (oof thats wordy)
- "Phantom type parameters can only appear as type arguments for type parameters that were declared 'phantom'" (oof wordy again, but I think you get the idea?)

Reply:
Yeah, separating between directly used as a field type and as type argument is a great improvement. Then ideally the second case could also include the position of the parameter that is declared as non-phantom.